### PR TITLE
Refactor file saving/exporting

### LINKED ISSUE #972: Refactor Save File Dialog
This is part of the overall dialog refactoring #959 

- remove hardcoded `save-file` dialog in favor of reusable dialogs
- implement new dialog in a way that it is configured through the client

### LINKED ISSUE #974: Refactor Close File Dialog
This is part of the overall dialog refactoring #959 

- remove hardcoded `close-file` dialog in favor of reusable dialogs
- implement new dialog in a way that it is configured through the client

### LINKED ISSUE #978: Refactor Export As Dialog
This is part of the overall dialog refactoring #959 

- remove hardcoded `export-as` dialog in favor of reusable dialogs
- implement new dialog in a way that it is configured through the client

### PR DESCRIPTION
**Pull request by philippfromme**
* refactor app/lib/dialog.js
* refactor app/lib/file-system.js
* refactor use of Dialog and FileSystem in app/lib/index.js
* refactor client/src/remote/Dialog.js
* refactor client/src/remote/FileSystem.js
* refactor saving and exporting tabs in client/src/app/App.js

Closes #972 
Closes #974 
Closes #978

### CODE REVIEW

**Comment by barmac** [comment]
There is a problem when using `Save as...` for DMN diagram. I get this error:

```objc[14230]: Class FIFinderSyncExtensionHost is implemented in both /System/Library/PrivateFrameworks/FinderKit.framework/Versions/A/FinderKit (0x7fff9bb551d0) and /System/Library/PrivateFrameworks/FileProvider.framework/OverrideBundles/FinderSyncCollaborationFileProviderOverride.bundle/Contents/MacOS/FinderSyncCollaborationFileProviderOverride (0x11e594dc8). One of the two will be used. Which one is undefined.
[workspace] save
[workspace] save
(node:14230) UnhandledPromiseRejectionWarning: TypeError: path.dirname is not a function
    at /Users/maciej.barelkowski/workspace/bpmn-io/camunda-modeler/app/lib/index.js:193:25
    at EventEmitter.<anonymous> (/Users/maciej.barelkowski/workspace/bpmn-io/camunda-modeler/app/lib/util/renderer.js:26:14)
    at EventEmitter.emit (events.js:182:13)
    at WebContents.<anonymous> (/Users/maciej.barelkowski/workspace/bpmn-io/camunda-modeler/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:284:13)
    at WebContents.emit (events.js:182:13)
(node:14230) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:14230) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:14230) UnhandledPromiseRejectionWarning: TypeError: path.dirname is not a function
    at /Users/maciej.barelkowski/workspace/bpmn-io/camunda-modeler/app/lib/index.js:193:25
    at EventEmitter.<anonymous> (/Users/maciej.barelkowski/workspace/bpmn-io/camunda-modeler/app/lib/util/renderer.js:26:14)
    at EventEmitter.emit (events.js:182:13)
    at WebContents.<anonymous> (/Users/maciej.barelkowski/workspace/bpmn-io/camunda-modeler/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:284:13)
    at WebContents.emit (events.js:182:13)
(node:14230) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)```

Steps to reproduce:
1. Open Camunda Modeler
2. Open a DMN diagram
3. Click `Save as...`

Nothing happens in the window, but the main process console shows the errors.

File: app/lib/index.js [hunk 1]
```diff
@@ -173,6 +167,41 @@ renderer.on('dialog:content-changed', function(done) {
...
+
+  const filePaths = await dialog.showOpenDialog(options);
+
+  done(null, filePaths);
+});
+
+renderer.on('dialog:save-file', async function(options, done) {
+  const { file } = options;
+
+  const { path } = file;
```
Comment (barmac) [inline app/lib/index.js]: This assignment shadows the `path` module. Easy fix for that is to use something like:

```javascript
const { path: filePath } = file;

if (filePath) {
  ...
```

**Comment by barmac** [comment]
I managed to identify the issue and pushed a fix for it :)

File: app/lib/index.js [hunk 2]
```diff
@@ -199,78 +228,58 @@ renderer.on('deploy', function(data, done) {
...
-    assign(options, {
-      defaultPath: path.dirname(activeFile.path)
-    });
-  }
-
-  const filePaths = await dialog.showOpenDialog(options);
+renderer.on('client-config:get', function(...args) {
 
-  const files = fileSystem.open(filePaths);
+  var done = args[args.length - 1];
```
Comment (barmac) [inline app/lib/index.js]: If `args[args.length -1]` is a callback, then we call `clientConfig#get` with that callback. Is it intended? Also, do you think we should we keep this structure with callback used for reporting errors only?

**Comment by philippfromme** [comment]
💪

File: app/lib/index.js (hunk 2 above)
Comment (philippfromme) [reply app/lib/index.js]: We'll take care of that in https://github.com/camunda/camunda-modeler/issues/1001

**Pull request merged**